Adjoint differentation improvements

albi3ro opened the pull request:
The [current adjoint differentiation](https://github.com/PennyLaneAI/pennylane/blob/98a787643f4928b6078fede6d4ecb7adab3e1800/pennylane/_qubit_device.py#L785) performs a forward pass of the circuit, even though PennyLane will often have performed one just before calling the adjoint jacobian function.  

This PR intends to make it easier to reduce the number of circuit executions when using adjoint diff.  Instead of using a "one-size-fits-all" caching method, this PR introduces several different ways users could reduce forward pass executions:

1) compute expectation values and the jacobian at the same time inside of `adjoint_jacobian` with the `return_obs` keyword

2) start with the current device state if requested by the `use_device_state` keyword

3) start with the provided state if it is provided to the `starting_state` keyword

A note that `starting_state` takes precedence over `use_device_state`.

Reducing the forward pass can reduce execution time of the device method call itself by a full 25%. 

I am also working on editing the interfaces to take advantage of this speed up.

josh146 commented:
🎉

albi3ro commented:
As of now, this branch is seeing decent time improvements.  I changed the benchmark repo's `GradientComputation_light` benchmark to use the adjoint method.  

The results for current master:
```
[  0.00%] · For pennylane commit 411c3d52 <master>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-Qulacs-jax-jaxlib-networkx-qsimcirq-tensorflow-torch
[  0.05%] ··· Running (asv.core_suite.GradientComputation_light.time_gradient--).
[  0.09%] ··· asv.core_suite.GradientComputation_light.time_gradient                                                                                       4/16 failed
[  0.09%] ··· ========= ============== ============ ============ ========= ============== ============ ============= =========
              --                                                 n_layers / interface
              --------- ------------------------------------------------------------------------------------------------------
               n_wires   3 / autograd     3 / tf     3 / torch    3 / jax   6 / autograd     6 / tf      6 / torch    6 / jax
              ========= ============== ============ ============ ========= ============== ============ ============= =========
                  2       5.11±0.1ms     10.4±2ms    4.77±0.1ms    failed    7.45±0.3ms    14.2±0.2ms   7.14±0.05ms    failed
                  5       10.2±0.2ms    18.6±0.5ms   9.76±0.3ms    failed    17.2±0.3ms    31.7±0.3ms    16.6±0.6ms    failed
              ========= ============== ============ ============ ========= ============== ============ ============= =========
```

And for the most recent commit to this PR:
```
    [  0.00%] · For pennylane commit 7a26a6ce <adjoint_diff_improvements>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-Qulacs-jax-jaxlib-networkx-qsimcirq-tensorflow-torch
[  0.05%] ··· Running (asv.core_suite.GradientComputation_light.time_gradient--).
[  0.09%] ··· asv.core_suite.GradientComputation_light.time_gradient                                                                                       3/16 failed
[  0.09%] ··· ========= ============== ============ ============ ========= ============== ============ ============ =========
              --                                                 n_layers / interface
              --------- -----------------------------------------------------------------------------------------------------
               n_wires   3 / autograd     3 / tf     3 / torch    3 / jax   6 / autograd     6 / tf     6 / torch    6 / jax
              ========= ============== ============ ============ ========= ============== ============ ============ =========
                  2      4.35±0.05ms    9.00±0.3ms    4.98±1ms    235±2ms    6.76±0.3ms     14.5±3ms    6.74±0.1ms    failed
                  5      8.63±0.03ms    17.0±0.6ms   9.82±0.2ms    failed    15.2±0.5ms    29.7±0.7ms   16.1±0.4ms    failed
              ========= ============== ============ ============ ========= ============== ============ ============ =========
```

albi3ro commented on pennylane/_qubit_device.py:
```diff
@@ -817,9 +817,17 @@ def adjoint_jacobian(self, tape):
         Args:
             tape (.QuantumTape): circuit that the function takes the gradient of
 
+        Kwargs:
```
```suggestion
        Keyword Args:
```

albi3ro commented:
@dwierichs want to help review?

trbromley commented on pennylane/interfaces/jax.py:
```diff
@@ -95,7 +95,9 @@ def _execute(self, params, device):
             exec_fn = partial(self.execute_device, device=device)
             return host_callback.call(
-                exec_fn, params, result_shape=jax.ShapeDtypeStruct((1,), JAXInterface.dtype)
+                exec_fn,
+                params,
+                result_shape=jax.ShapeDtypeStruct((1,), JAXInterface.dtype),
```
Is JAX still a TODO, or it already skips the extra first pass?

albi3ro replied:
JAX and autograd use a keyword set in qnode initialization.  I had been setting it here but thought it better to add to the validation step.

trbromley commented on pennylane/interfaces/tf.py:
```diff
@@ -138,6 +138,11 @@ def _execute(self, params, **input_kwargs):
         res = self.execute_device(args, input_kwargs["device"])
         self.set_parameters(all_params, trainable_only=False)
 
+        # save state
+        jac_options = getattr(self, "jacobian_options", dict())
+        if jac_options.get("cache_state", False):
+            state = input_kwargs["device"].state
+
```
```suggestion
        # tape might not be a jacobian tape
        jac_options = getattr(self, "jacobian_options", dict())
        # cache state for adjoint jacobian computation
        if jac_options.get("cache_state", False):
            state = input_kwargs["device"].state

```
To match the torch style

trbromley commented on pennylane/interfaces/tf.py:
```diff
@@ -174,6 +179,9 @@ def _evaluate_grad_matrix(grad_matrix_fn):
                 return saved_grad_matrices[grad_matrix_fn]
 
+            if self.jacobian_options.get("cache_state", False):
+                self.jacobian_options["device_pd_options"] = {"starting_state": state}
```
Could we just have this line within the `if jac_options.get("cache_state", False):` statement above? Is there a reason to do it within `_evaluate_grad_matrix`?

albi3ro replied:
`_evaluate_grad_matrix` could be called later on after arbitrary other executions on the device.  This function `_evaluate_grad_matrix` will remember the value of state where the function was constructed.  It basically caries around the state with it.

I don't want to set the tape keyword till the last moment before taking the derivative in case it gets overwritten by a different execution.

albi3ro replied:
This keyword gets set during qnode initialization time.  Directly manipulating tape jacobian options isn't particularly user-facing, but I don't think an overabundance of caution here will hurt anyone.

trbromley commented on pennylane/qnode.py:
```diff
@@ -123,6 +123,9 @@ class QNode:
         order=1 (int): The order of the finite difference method to use. ``1`` corresponds
             to forward finite differences, ``2`` to centered finite differences.
         shift=pi/2 (float): the size of the shift for two-term parameter-shift gradient computations
+        cache_state=False (bool): for tensorflow and torch interfaces and adjoint differentiation,
+            this indicates whether to save the device state after the forward pass.  Doing so saves a
+            forward execution. Device state automatically reused with autograd and jax interfaces.
```
Is there a reason not to have this on when we're doing adjoint differentiation? If we always want it on, then maybe we don't have this keyword argument and just have the state always cached (provided that adjoint diff method is requested)

albi3ro replied:
Given state's might occupy a bit of memory, I don't want the memory requirements to balloon.  We may make `True` the default later, but I thought it best to leave it `False` to begin with.

trbromley replied:
:thinking: I was on the fence about adding `cache_state` as a keyword argument, feeling that it might pollute the keyword arguments of the QNode for a fairly specific use case. However, can agree that it might be nice to have the option to turn off/on this feature due to memory limitations. One suggestion would be to rename to something more specific, e.g., `adjoint_cache` since perhaps `cache_state` might give the impression of something more general than it is.

I'd also suggest making `cache_state=True` by default. If it's off by default, I think most users will not notice and not benefit from the feature, so all the hard work in this PR will not be quite so beneficial. Any users experiencing memory issues can always turn off caching if they have a problem.

trbromley replied:
Actually, now I look again, when `cache_state=True`, we cache the state in the Torch and TF interfaces regardless of the `diff_method`? :thinking:

Shouldn't we have it so that caching only occurs if `diff_method="adjoint"`? We can expand to further use cases going forward.

In other words,
Currently: off by default, but when on caching occurs in TF and Torch interfaces regardless of diff method
My suggestion: on by default, and caching occurs in TF and Torch interfaces only when diff method is adjoint

I understand the caution with off-by-default in the former, but I think it's worth following a smart defaults philosophy.

trbromley commented on tests/interfaces/test_qnode_jax.py:
```diff
@@ -133,6 +133,25 @@ def circuit(a, p):
 
 
+def test_adjoint_reuse_device_state(mocker):
+    """Tests that the autograd interface reuses the device state for adjoint differentiation"""
```
```suggestion
    """Tests that the jax interface reuses the device state for adjoint differentiation"""
```

trbromley commented on pennylane/_qubit_device.py:
```diff
@@ -873,20 +884,24 @@ def adjoint_jacobian(self, tape):
+            bras = [self._apply_operation(bra_, op) for bra_ in bras]
             op.inv()
 
+        if return_obs:
+            return expectation, jac
```
What is the usecase for `return_obs`? Maybe I missed it, but it doesn't seem to be being used by any of the interfaces.

albi3ro replied:
It's not currently used by the interfaces. I wanted to make this function more versatile.  Even though it's not currently used elsewhere in the code base, I think it makes the method itself stronger.

trbromley commented on pennylane/qnode.py:
```diff
@@ -411,11 +414,17 @@ def _validate_adjoint_method(device, interface):
 
+        jac_options = {"method": "device", "jacobian_method": "adjoint_jacobian"}
+        # reuse the forward pass
+        # torch and tensorflow can cache the state
+        if interface in {"autograd", "jax"}:
+            jac_options["device_pd_options"] = {"use_device_state": True}
```
I'm confused on the difference between the `use_device_state` workflow and the `starting_state` workflow.

So we have:
- Autograd and jax use the device state
- Torch and TF use `starting_state`, where `starting_state` is the device's state but loaded from the forward pass within the interface

Can autograd and jax not use `starting_state`? Can Torch and TF not use the device state approach?

albi3ro replied:
I don't want autograd and jax to use `starting_state`, as that requires more moving around of large arrays.  Torch and TF can't use the device state, because we might perform the backward pass after multiple forward passes.

trbromley reviewed (changes requested):
Thanks @albi3ro! Looks good, my main question being about the difference between the `starting_state` and `use_device_state` approaches for each interface - I guess we can't just roll with one option?

albi3ro commented:
@trbromley  For the autograd and jax derivatives, we know that we just performed a forward pass, so the device is in the correct state.

For the tensorflow and torch interfaces, we can perform the forward pass, the go do other stuff, maybe perform a forward pass with different parameters, and *then* perform the backward pass where we perform `dev.adjoint_jacobian(qtape)`.  So we cannot use the current device state, since other things may have occurred on the device since the forward pass.

Caching the device state means keeping around a copy of the statevector, which may come with some memory overhead.  At some point, we may switch to having this as default, but I don't want to do so just yet given the memory considerations.

albi3ro commented on pennylane/_qubit_device.py:
```diff
@@ -836,16 +844,19 @@ def adjoint_jacobian(self, tape):
-        # https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
-        self.reset()
-        self.execute(tape)
+        if starting_state is not None:
+            ket = self._reshape(starting_state, [2] * self.num_wires)
```
I wanted to make it so that someone could do

```
state = dev.state
dev.adjoint_jacobian(qtape, starting_state=state)
```

`dev.state` is flattened, so I added this reshape to accommodate doing the above.

albi3ro replied:
doing so.

albi3ro replied:
Actually, I just realized that reshaping is an incredibly efficient operation because it is a different view of the same data.  And this decreases readability slightly.  So I'm always going to reshape starting state.

dwierichs replied:
Just ran some example timing and had to laugh at how negligible the cost of reshaping are! :D Nice!

dwierichs reviewed (commented):
Very cool improvements, just left very few comments.
Just to make sure, as my experience with the interfaces is quite limited still: Is there absolutely no way one could break the caching with intermediate calls / changes of the device/qnode state in the sense of recycling the wrong state?

albi3ro commented on pennylane/_qubit_device.py:
```diff
@@ -836,16 +844,19 @@ def adjoint_jacobian(self, tape):
+        if starting_state is not None:
+            ket = self._reshape(starting_state, [2] * self.num_wires)
+        elif use_device_state:
+            ket = self._reshape(self.state, [2] * self.num_wires)
```
Substituting for `self._prerotated_state`

albi3ro commented on pennylane/_qubit_device.py:
```diff
@@ -873,20 +884,24 @@ def adjoint_jacobian(self, tape):
-                    jac[:, trainable_param_number] = jac_column
+                    ket_temp = self._apply_unitary(ket, d_op_matrix, op.wires)
+
+                    for kk, bra_ in enumerate(bras):
+                        jac[kk, trainable_param_number] = 2 * dot_product_real(bra_, ket_temp)
```
`bras` is a list of arrays, rather than an array with an extra dimension.  I thought about making it an array with an additional dimension.  Think I'll go ahead and proceed with trying to make that work.

dwierichs replied:
Ah, yes, sorry I overlooked that! The multiplication should be saving more than the conversion to array costs, I'd hope :)

albi3ro commented:
```
[  0.00%] · For pennylane commit d8b6011d <adjoint_diff_improvements>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-Qulacs-jax-jaxlib-networkx-qsimcirq-tensorflow-torch
[  0.04%] ··· Running (asv.core_suite.GradientComputation_light.time_gradient--).
[  0.09%] ··· ...uite.GradientComputation_light.time_gradient        4/16 failed
[  0.09%] ··· ========= ========== =========== =============
               n_wires   n_layers   interface
              --------- ---------- ----------- -------------
                  2         3        autograd   4.37±0.08ms
                  2         3           tf       8.90±0.1ms
                  2         3         torch     4.68±0.09ms
                  2         3          jax         failed
                  2         6        autograd    6.54±0.2ms
                  2         6           tf       14.0±0.1ms
                  2         6         torch      6.83±0.1ms
                  2         6          jax         failed
                  5         3        autograd    8.89±0.3ms
                  5         3           tf       17.9±0.4ms
                  5         3         torch      9.45±0.1ms
                  5         3          jax         failed
                  5         6        autograd    15.7±0.2ms
                  5         6           tf       31.6±0.1ms
                  5         6         torch      16.5±0.1ms
                  5         6          jax         failed
              ========= ========== =========== =============


[  0.00%] · For pennylane commit 0dcbed53 <master>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-Qulacs-jax-jaxlib-networkx-qsimcirq-tensorflow-torch
[  0.04%] ··· Running (asv.core_suite.GradientComputation_light.time_gradient--).
[  0.09%] ··· ...uite.GradientComputation_light.time_gradient        3/16 failed
[  0.09%] ··· ========= ========== =========== =============
               n_wires   n_layers   interface
              --------- ---------- ----------- -------------
                  2         3        autograd   4.81±0.05ms
                  2         3           tf       9.14±0.1ms
                  2         3         torch     4.60±0.04ms
                  2         3          jax         failed
                  2         6        autograd    7.58±0.3ms
                  2         6           tf       13.9±0.2ms
                  2         6         torch      6.98±0.3ms
                  2         6          jax         failed
                  5         3        autograd   10.1±0.09ms
                  5         3           tf       18.1±0.3ms
                  5         3         torch      9.81±0.3ms
                  5         3          jax        682±8ms
                  5         6        autograd    17.6±0.3ms
                  5         6           tf       31.7±0.1ms
                  5         6         torch     16.4±0.05ms
                  5         6          jax         failed
              ========= ========== =========== =============
```

Looks like some modest gains in autograd, but not much change for tensorflow and torch.  Jax, as before, is problematic when getting timed. I'm curious why tensorflow and torch don't show a difference.

trbromley commented on tests/interfaces/test_qnode_tf.py:
```diff
@@ -780,6 +787,70 @@ def circuit(x):
+class Test_adjoint:
+    def test_adjoint_save_state(self, mocker):
+        """Tests that the tf interface reuses device state when prompted by `cache_state=True`.
+        Also makes sure executing a second circuit before backward pass does not interfere
+        with answer.
```
Nice to test this edge case!

trbromley commented on tests/interfaces/test_qnode_torch.py:
```diff
@@ -717,6 +736,67 @@ def circuit(x):
+class Test_adjoint:
+    def test_adjoint_save_state(self, mocker):
+        """Tests that the torch interface reuses device state when prompted by `cache_state=True`.
+
```
```suggestion
```

trbromley commented on tests/test_qubit_device_adjoint_jacobian.py:
```diff
@@ -192,6 +192,64 @@ def test_gradient_gate_with_multiple_parameters(self, tol, dev):
         # the different methods agree
         assert np.allclose(grad_D, grad_F, atol=tol, rtol=0)
 
+    def test_return_expectation(self, tol, dev):
```
I'm still not sure about the addition of `return_obs` - do you have a future use-case in mind?

albi3ro replied:
Easy enough to add in later if a use case arises, so I'll remove it for now.

trbromley reviewed (commented):
Thanks @albi3ro! The code is on point and am happy to give a :heavy_check_mark:, but I just wanted to discuss more the question of `cache_state`. Concretely, my suggestion would be to:

- Make `cache_state` only cache when using the adjoint diff method
- Have it be on by default
- Rename to `adjoint_cache` or similar

Would be great to get your thoughts though!

albi3ro commented on pennylane/_qubit_device.py:
```diff
@@ -836,16 +856,23 @@ def adjoint_jacobian(self, tape):
 
-        phi = self._reshape(self.state, [2] * self.num_wires)
+        n_obs = len(tape.observables)
+        bras = np.empty([n_obs] + [2] * self.num_wires, dtype=np.complex128)
+        for kk in range(n_obs):
+            bras[kk, ...] = self._apply_operation(ket, tape.observables[kk])
 
-        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+        # this can probably be more optimized, but at least now it works...
+        if return_obs:
+            expectation = dot_product_real(bras, ket)
```
Yep

albi3ro commented:
@trbromley Good catch on realizing that tensorflow and torch would be catching the state even when not using adjoint diff!

Now the state gets cached by default, but only if using the adjoint diff method.

I've renamed the keyword to `adjoint_cache`.

I've also removed `return_obs`.  May be useful in the future at some point, but I can easily add it in when the use case arises.

trbromley commented on .github/CHANGELOG.md:
```diff
@@ -310,12 +310,15 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 <h3>Improvements</h3>
 
+
```
```suggestion
```

trbromley commented on .github/CHANGELOG.md:
```diff
@@ -310,12 +310,15 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
+
+* The adjoint jacobian differentiation method reuses the state computed on the forward pass.
+[(#1341)](https://github.com/PennyLaneAI/pennylane/pull/1341)
```
```suggestion
  [(#1341)](https://github.com/PennyLaneAI/pennylane/pull/1341)
```
Not sure if this spacing really matter :thinking:

trbromley commented on pennylane/qnode.py:
```diff
@@ -123,6 +123,9 @@ class QNode:
         order=1 (int): The order of the finite difference method to use. ``1`` corresponds
             to forward finite differences, ``2`` to centered finite differences.
         shift=pi/2 (float): the size of the shift for two-term parameter-shift gradient computations
+        adjoint_cache=True (bool): for tensorflow and torch interfaces and adjoint differentiation,
+            this indicates whether to save the device state after the forward pass.  Doing so saves a
+            forward execution. Device state automatically reused with autograd and jax interfaces.
```
```suggestion
        adjoint_cache=True (bool): for TensorFlow and PyTorch interfaces and adjoint differentiation,
            this indicates whether to save the device state after the forward pass.  Doing so saves a
            forward execution. Device state automatically reused with autograd and JAX interfaces.
```
Picky, but just to follow rest of docs.

trbromley commented on tests/interfaces/test_qnode_torch.py:
```diff
@@ -717,6 +736,66 @@ def circuit(x):
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev, diff_method="adjoint", interface="torch")
```
```suggestion
        @qml.qnode(dev, diff_method="adjoint", interface="torch", cache_state=False)
```

trbromley commented on tests/interfaces/test_qnode_tf.py:
```diff
@@ -780,6 +787,70 @@ def circuit(x):
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev, diff_method="adjoint", interface="tf")
```
```suggestion
        @qml.qnode(dev, diff_method="adjoint", interface="tf", cache_state=False)
```

trbromley commented on .github/CHANGELOG.md:
```diff
@@ -311,9 +311,7 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
-* The adjoint jacobian differentiation method reuses the state computed on the forward pass. For torch and tensorflow
-interfaces, this is disabled by default. By passing `cache_state=True` during QNode creation for tensorflow and 
-torch interfaces, the user can save the device state for the backward pass.
+* The adjoint jacobian differentiation method reuses the state computed on the forward pass.
```
```suggestion
* The adjoint jacobian differentiation method reuses the state computed on the forward pass.
  This can be turned off to save memory with the Torch and TensorFlow interfaces by passing
  `adjoint_cache=False` during QNode creation.
```
Worth mentioning?

trbromley commented on pennylane/qnode.py:
```diff
@@ -941,6 +950,9 @@ def qnode(
         h=1e-7 (float): Step size for the finite difference method.
         order=1 (int): The order of the finite difference method to use. ``1`` corresponds
             to forward finite differences, ``2`` to centered finite differences.
+        adjoint_cache=True (bool): for tensorflow and torch interfaces and adjoint differentiation,
+            this indicates whether to save the device state after the forward pass.  Doing so saves a
+            forward execution. Device state automatically reused with autograd and jax interfaces.
```
```suggestion
        adjoint_cache=True (bool): for TensorFlow and PyTorch interfaces and adjoint differentiation,
            this indicates whether to save the device state after the forward pass.  Doing so saves a
            forward execution. Device state automatically reused with autograd and JAX interfaces.
```

trbromley reviewed (approved):
Thanks @albi3ro, awesome!

trbromley commented on tests/interfaces/test_qnode_torch.py:
```diff
@@ -717,6 +736,66 @@ def circuit(x):
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, diff_method="adjoint", interface="torch", cache_state=True)
```
```suggestion
        @qml.qnode(dev, diff_method="adjoint", interface="torch", adjoint_cache=True)

```

trbromley commented on tests/interfaces/test_qnode_tf.py:
```diff
@@ -780,6 +787,70 @@ def circuit(x):
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, diff_method="adjoint", interface="tf", cache_state=True)
```
```suggestion
        @qml.qnode(dev, diff_method="adjoint", interface="tf", adjoint_cache=True)
```
